### ISSUE TITLE
`prop-types` rule reports for Array.prototype.reduce is missing

### ISSUE THREAD

**Issue by ivalduan** [yannickcr/eslint-plugin-react#2383]
Similar to #1047 when a propType React.PropTypes.arrayOf(<something>), you shouldn't be prompted to add reduce by eslint.
By extension any Array.prototype method that does not modify the data should be included.

**Comment by ivalduan** [comment]
Reproduced with v7.14.2

**Comment by ljharb** [comment]
Agreed, this is a bug

**Comment by golopot** [comment]
I cannot reproduce this. Can you share a snippet for reproduction?

```js
import React from 'react';
import PropTypes from 'prop-types';

function Foo({ a }) {
  return <p>{a.reduce(() => 5)}</p>;
}

Foo.propTypes = {
  a: PropTypes.arrayOf(PropTypes.string).isRequired,
};
```

**Comment by ivalduan** [comment]
Thought it was simple but it seems like the problem goes in another direction.

I was able to reproduce with the following snippet, it seems to be related with identifier collision when parsing the code, if the destructured variable is named differently than the root prop eslint recognizes well the array and the reduce.
```js
import React from 'react';
import PropTypes from 'prop-types';

function Foo({
  elements: {
    items: elements,
  },
}) {
  return <p>{elements.reduce(() => 5)}</p>;
}

Foo.propTypes = {
  elements: PropTypes.shape({
    items: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
};
```
